### PR TITLE
Use node-fetch v3 beta and require Node.js 10.16

### LINKED ISSUE #8: clone() hangs with large response in Node
[Issuehunt badges](https://issuehunt.io/r/sindresorhus/ky-universal/issues/8)

I hit this bug today, it was caused by https://github.com/bitinn/node-fetch/issues/386

The bug makes `ky-universal` not useable at all in node. (You don't know when the code will hang as you don't know the response size ahead of time)

Is there a way to remove `clone()` from the source code?

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Backers (Total: $80.00)

- [issuehunt issuehunt](https://issuehunt.io/u/issuehunt) ($80.00)

### Submitted pull Requests
- [#19 Use node-fetch v3 beta](https://issuehunt.io/r/sindresorhus/ky-universal/pull/19)
---

### Tips

- Checkout the [Issuehunt explorer](https://issuehunt.io/r/sindresorhus/ky-universal/) to discover more funded issues.
- Need some help from other developers? [Add your repositories](https://issuehunt.io/r/new) on IssueHunt to raise funds.
</details>

### PR DESCRIPTION
**Pull request by xxczaki**
Fixes #8 (allows using the `highWaterMark` option)

---

We have a [section](https://github.com/node-fetch/node-fetch#custom-highwatermark) in the node-fetch's readme about custom `highWaterMark`. Let me know if we should link it somewhere here :smile:

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Referenced issues

This pull request has been submitted to:
- [#8: clone() hangs with large response in Node](https://issuehunt.io/repos/172099953/issues/8)
---
</details>

### CODE REVIEW

**Comment by xxczaki** [comment]
Tests are failing on Node.js v8, as [it is not supported in the `v3` version of node-fetch](https://github.com/node-fetch/node-fetch/blob/master/docs/v3-UPGRADE-GUIDE.md#minimum-supported-nodejs-version-is-now-10).

**Comment by sholladay** [comment]
Thanks for the PR!

I am 👍 on dropping Node 8.

**Comment by xxczaki** [comment]
@sholladay I added information about how to deal with `highWaterMark` to FAQ.

**Comment by sholladay** [comment]
Could you double check the unit measurements? I would expect these values to be measured in multiples of bytes, not bits.

**Comment by xxczaki** [comment]
Oh, you are right. Just changed it. Thanks for catching that!

File: readme.md [hunk 1]
```diff
@@ -95,6 +95,25 @@ Put the following in package.json:
...
+
+One way to fix this is to specify the `highWaterMark`:
+
+```js
+import ky from 'ky-universal';
+
+(async () => {
+	const response = await ky('https://example.com', {
+		// About 1MB
+		highWaterMark: 1024 * 1024
```
Comment (sindresorhus) [inline readme.md:110]: Why can't we just set a reasonable number in `ky-universal` unless it's set by the user? That way we would fix it for most use-cases.

Comment (paambaati) [reply readme.md:110]: +1 for this. This would help avoid an issue that a good number of users would encounter anyway.

Comment (xxczaki) [reply readme.md:110]: @sindresorhus If so, is `1024 * 1024` ok or do you want to use a different value?

Comment (paambaati) [reply readme.md:110]: FWIW, the [default for the `fs` module is 64kb](https://nodejs.org/api/fs.html#fs_fs_createreadstream_path_options) instead of 16kb like streams.

Comment (xxczaki) [reply readme.md:110]: @paambaati Node Fetch does not use `fs.createReadStream`.

Comment (sholladay) [reply readme.md:110]: I think we should aim to handle some common use cases by default, such as downloading a large image. 10MB feels like a comfortable value to me. We should err on the high side, because the clone just hangs when the buffer becomes full.

Comment (paambaati) [reply readme.md:110]: Would it be a good idea (if even possible) to print a warning and a recommendation to bump up `highWatermark` when the buffer gets full?

Comment (paambaati) [reply readme.md:110]: Would it be a good idea (if even possible) to print a warning and a recommendation to bump up `highWatermark` when the buffer gets full?

Comment (sholladay) [reply readme.md:110]: Unless `node-fetch` has some sort of event hook for its internal buffer, that sounds tricky, because we sometimes clone the response even when we are not using streams. At the end of the day, I'm pretty sure setting `highWaterMark` is just a workaround and the buffering implementation in `node-fetch` needs to be fixed.

Comment (sindresorhus) [reply readme.md:110]: 10 MB sounds good to me.

**Comment by xxczaki** [comment]
The default `highWaterMark` is now 10MB. Should we mention it in the docs or not?

File: readme.md [hunk 2]
```diff
@@ -95,6 +95,25 @@ Put the following in package.json:
...
+Stream in Node.js have a smaller internal buffer size (16kB, aka `highWaterMark`) from client-side browsers (>1MB, not consistent across browsers). Because of that, when you are writing an isomorphic app and using `res.clone()`, it will hang with large response in Node.
+
+One way to fix this is to specify the `highWaterMark`:
+
+```js
+import ky from 'ky-universal';
+
+(async () => {
+	const response = await ky('https://example.com', {
+		// About 1MB
```
Comment (sindresorhus) [inline readme.md:109]: This needs to be higher than 10 MB now. Otherwise, the example is pointless.

**Comment by sindresorhus** [comment]
> Should I mention it in the docs or not?

Yes

File: readme.md (hunk 2 above)
Comment (xxczaki) [reply readme.md]: Done.

**Comment by xxczaki** [comment]
Everything done :smile:

File: index.js [hunk 3]
```diff
@@ -3,7 +3,7 @@ const fetch = require('node-fetch');
 const AbortController = require('abort-controller');
 
 if (!global.fetch) {
-	global.fetch = fetch;
+	global.fetch = (url, options) => fetch(url, {options, ...{highWaterMark: 10240 * 1024}});
```
Comment (sholladay) [inline index.js:6]: This doesn't look right. It's assigning the options to a property named `options`. That doesn't work, does it?

Comment (xxczaki) [reply index.js]: 🤦

Comment (xxczaki) [reply index.js]: @sholladay Sorry for that, just fixed it.

File: index.js [hunk 4]
```diff
@@ -2,8 +2,10 @@
 const fetch = require('node-fetch');
 const AbortController = require('abort-controller');
 
+const TEN_MEGABYTES = 1000 * 1000 * 10;
+
 if (!global.fetch) {
-	global.fetch = fetch;
+	global.fetch = (url, options) => fetch(url, {...options, ...{highWaterMark: TEN_MEGABYTES}});
```
Comment (sholladay) [inline index.js:8]: This is overriding the user's options. Meaning, they can't set a custom `highWaterMark`.

File: index.js [hunk 5]
```diff
@@ -3,7 +3,9 @@ const fetch = require('node-fetch');
 const AbortController = require('abort-controller');
 
 if (!global.fetch) {
-	global.fetch = (url, options) => fetch(url, {options, ...{highWaterMark: 10240 * 1024}});
+	global.fetch = (url, options) => {
+		return fetch(url, {...options, ...{highWaterMark: 10240 * 1024}});
```
[the remark this answers is not in the document]
Comment (sindresorhus) [reply index.js]:
```
fetch(url, {highWaterMark: 10240 * 1024, ...options});
```

File: index.js [hunk 6]
```diff
@@ -2,8 +2,10 @@
 const fetch = require('node-fetch');
 const AbortController = require('abort-controller');
 
+const TEN_MEGABYTES = 10240 * 1024;
+
 if (!global.fetch) {
-	global.fetch = fetch;
+	global.fetch = options => fetch(options.url, Object.assign(options, {highWaterMark: options.highWaterMark || TEN_MEGABYTES}));
```
Comment (xxczaki) [inline index.js:8]: @sholladay I'm not sure whether there is a cleaner way to write this, but I can confirm it works now:

```js
test('default highWaterMark is 10MB', async t => {
	const response = await ky('https://httpbin.org/json');
	t.is(response.highWaterMark, 10485760);
});

test('can override the highWaterMark', async t => {
	const response = await ky('https://httpbin.org/json', {highWaterMark: 1024 * 1024});
	t.is(response.highWaterMark, 1048576);
});
```

Comment (sholladay) [reply index.js:8]: I'm away from my laptop at the moment, but I'm happy to help tomorrow. Basically, it should use object spread the way it used to be in your previous commit, but the default options need to come before the user's options object. When using object spread, the order matters, with later options overriding earlier ones. Does that make sense?

Comment (paambaati) [reply index.js:8]: @xxczaki Basically this -

```suggestion
	global.fetch = (url, options) => fetch(url, {highWaterMark: TEN_MEGABYTES, ...options});
```

Comment (xxczaki) [reply index.js:8]: @sholladay The problem is, in `node-fetch@3.0.0-beta.5`, the default `highWaterMark` is undefined and not 16384. This was fixed and therefore we will be able to use the spread syntax with `3.0.0-beta.6`, which should come out in a few days max. 

So I think we should wait for the next release ;)

File: index.js [hunk 7]
```diff
@@ -2,10 +2,10 @@
 const fetch = require('node-fetch');
 const AbortController = require('abort-controller');
 
-const TEN_MEGABYTES = 1000 * 1000 * 10;
+const TEN_MEGABYTES = 10240 * 1024;
```
Comment (sindresorhus) [inline index.js:5]: Don't change what I had there. This is both not readable and not correct. What you have there is 10 MiB, not 10 MB. We'll do 10 MB.

**Comment by xxczaki** [comment]
I updated the node-fetch version so that we can now use the spread syntax. However, it will not work the way you wanted to:
```js
// Updates the default 16384 bytes, but does not allow the user to specify the `highWaterMark`
fetch(options.url, {...options, highWaterMark: TEN_MEGABYTES});
```
That's why I used the following approach:
```js
const DEFAULT = 16384;
fetch(options.url, {...options, highWaterMark: options.highWaterMark === DEFAULT ? TEN_MEGABYTES : options.highWaterMark});
```
Let me know if there is a cleaner way to do it.

**Comment by sindresorhus** [comment]
How about:

```js
fetch(options.url, {...options, ...{highWaterMark: TEN_MEGABYTES}});
```

?

**Comment by sholladay** [comment]
I changed it to the only thing that really makes sense to me:

```js
global.fetch = (url, options) => fetch(url, {highWaterMark: TEN_MEGABYTES, ...options});
```

Could you give this a try and let me know if it works? If not, I'll need to dig into it further, as the older commit with `DEFAULT` doesn't really make sense to me.

**Comment by xxczaki** [comment]
@sindresorhus This changes the default node-fetch's `highWaterMark` to 10 MB, but does not allow users to override it (so the same behavior as `{...options, highWaterMark: TEN_MEGABYTES}`).

@sholladay This doesn't change the default node-fetch's `highWaterMark` (16384 bytes) but does allow users to override it.

Here is the set of tests I use to check the behavior:

```js
test('default highWaterMark is 10MB', async t => {
	const response = await ky('https://httpbin.org/json');
	t.is(response.highWaterMark, 10000000);
});

test('can override the highWaterMark', async t => {
	const response = await ky('https://httpbin.org/json', {highWaterMark: 1024 * 1024});
	t.is(response.highWaterMark, 1048576);
});
```

In terms of the old solution (with `DEFAULT`) - it basically sets the `highWaterMark` to 10MB if `options.highWaterMark` equals 16384 bytes (node-fetch's default) and otherwise sets it to `options.highWaterMark` (it was overwritten by the user).

I hope that makes sense.

**Comment by xxczaki** [comment]
In conclusion, `fetch(options.url, {highWaterMark: TEN_MEGABYTES, ...options})` doesn't work, because we are setting the `highWaterMark` to 10MB, while at the same time ignoring the user preference (`options.highWaterMark`).

**Comment by sholladay** [comment]
How is `response.highWaterMark` being defined in your tests? It's a request option, I'm not sure how it ends up on the response.

> @sholladay This doesn't change the default node-fetch's highWaterMark (16384 bytes) but does allow users to override it.

I don't think we care about `node-fetch`'s default, we just want to have our own default, which is what my code does. I could totally be misunderstanding something, just not seeing it yet. Under what circumstances would our default not be applied, other than the user overriding it? And how/why?

> Also, we can't do (url, options) => ..., as the options will always return undefined.

This makes me think there might be a misunderstanding in how object spread works. Here's an example with mock fetch functions that take the same arguments and spread the options the same way, which do produce the output I would expect...

```js
const fetchA = (url, options) => {
    fetchB(url, {highWaterMark: 'fetchA default', ...options});
};
const fetchB = (url, options) => {
    console.log('url:', url);
    console.log('options:', options);
};

fetchA();
// url: undefined
// options: { highWaterMark: "fetchA default" }
fetchA('foo');
// url: "foo"
// options: { highWaterMark: "fetchA default" }
fetchA('foo', {});
// url: "foo"
// options: { highWaterMark: "fetchA default" }
fetchA('foo', {highWaterMark: 'custom value'})
// url: "foo"
// options: { highWaterMark: "custom value" }
```

In other words, it's fine to spread `undefined` into an object. The object will still be created, the default will still apply, and from `fetchB`'s perspective, there will always be an options object with a `highWaterMark`, with either the `fetchA` default, or the user's custom value. In either case, any default inside of `fetchB` should be ignored.

In this case, `fetchA` represents `ky-universal` and `fetchB` represents `node-fetch`.

Could you point me more specifically to where you think this logic is wrong?

**Comment by xxczaki** [comment]
Ok, it turns out I did not understand some things correctly :laughing: After doing more verification, I can confirm that the correct `highWaterMark` is applied in both cases.

**Comment by sholladay** [comment]
Great, no worries! This is looking good to me now. :+1: 

My final thought on this is that, ultimately, the `node-fetch` bug is still there, the stream could still hang if the response is large enough. Looking at their updated docs, I see that they recommend resolving both the original and cloned response at the same time and seem to imply that is the true, proper solution. I'm not sure if we can do that in all cases in Ky, but I think it should be possible in at least one or two of the cases where we clone the response. So that's something to look into for the future.
